woocommerce/google-listings-and-ads
# Enable filtering of WC REST API responses based on the gla_syncable query parameter

### Changes proposed in this Pull Request:



Closes part of https://github.com/woocommerce/google-listings-and-ads/issues/2146

This PR allows the filtering of private metadata (metadata prefixed with `_`), excluding metadata with the prefix `_wc_gla`, when a WC REST API request is made using a specific parameter, `gla_syncable`.

Additionally, this PR ensures that only the products the merchant intends to sync are returned. In this case, the required metadata setting is: `_wc_gla_visibility` = `sync-and-show`.

The WC REST API endpoints that need to be filter are the following:

| Endpoint  | Filters |
| ------------- | ------------- |
| `/wp-json/wc/v3/products`  | private metadata + _wc_gla_visibility = sync-and-show  |
| `/wp-json/wc/v3/products/<product_id>`l  | private metadata + _wc_gla_visibility = sync-and-show  |
| `/wp-json/wc/v3/products/<product_id>/variations`  | private metadata  |
| `/wp-json/wc/v3/products/<product_id>/variations/<variation_id>`l  | private metadata  |
| `/wp-json/wc/v3/coupons`l  | private metadata + _wc_gla_visibility = sync-and-show + email restrictions |
| `/wp-json/wc/v3/coupons/<coupon_id>`l  | private metadata + _wc_gla_visibility = sync-and-show + email restrictions  |

Filtering is unnecessary for shipping zone endpoints as they do not contain metadata.

- See here for more context: pcTzPl-1Qd-p2#comment-2663
- Here a list of all endponts: pcTzPl-1Qd-p2#comment-2650

### Detailed test instructions:




1. Make requests to the endpoints mentioned above and add the `gla_syncable` param, for example: `/wp-json/wc/v3/products?gla_syncable=1`
2. See that the data is filtered correctly.
3. If you remove the query param: `gla_syncable` it should not filter any data. 


### Additional details:

- Variants are special cases since they don't rely on `_wc_gla_visibility` and instead depend on the parent attribute. Hence, if we aim to avoid fetching a variant that the parent product doesn't have the `gla_visibility`, we'd have to check its parent, meaning an additional query. I'm unsure if it's necessary to go to that far, so I'll seek feedback in the project thread.
- Coupons with email restrictions should also be filtered:

![image](https://github.com/woocommerce/google-listings-and-ads/assets/2488994/42dd8ffc-4357-43fc-8abe-6d8e4be4ed88)



### Changelog entry

> Add - Filter WC REST API responses for gla_syncable param
